pam/gdm/protocol_test.go

```diff
@@ -0,0 +1,672 @@
+			wantError: "missing request type",
+		},
+
+		"Error request packet with unexpected data": {
+			gdmData: &Data{
+				Type: DataType_request,
+				Request: &RequestData{
+					Type: RequestType_changeStage,
+					Data: &RequestData_ChangeStage{},
+				},
+				Event: &EventData{},
+			},
+
+			wantError: "field Event should not be defined",
+		},
+
+		"Error response packet with missing data": {
+			gdmData: &Data{Type: DataType_response},
+
+			wantError: "missing response data",
+		},
+
+		"Error response packet with missing type": {
+			gdmData: &Data{
+				Type:     DataType_response,
+				Response: &ResponseData{Data: &ResponseData_Ack{}},
+			},
+
+			wantError: "missing response type",
+		},
+
+		"Error response packet with invalid type": {
+			gdmData: &Data{
+				Type:     DataType_response,
+				Response: &ResponseData{Type: RequestType(-1), Data: &ResponseData_Ack{}},
+			},
+
+			wantError: "unexpected request type -1",
+		},
+
+		"Error response packet with unexpected data": {
+			gdmData: &Data{
+				Type:     DataType_response,
+				Response: &ResponseData{Type: RequestType_changeStage, Data: &ResponseData_Ack{}},
+				Event:    &EventData{},
+			},
+
+			wantError: "field Event should not be defined",
+		},
+
+		"Error poll packet with unexpected data": {
+			gdmData: &Data{Type: DataType_poll, Request: &RequestData{}},
+
+			wantError: "field Request should not be defined",
+		},
+
+		"Error pollResponse packet with missing event type": {
+			gdmData: &Data{
+				Type: DataType_pollResponse,
+				PollResponse: []*EventData{
+					{
+						Type: EventType_brokerSelected,
+						Data: &EventData_BrokerSelected{
+							BrokerSelected: &Events_BrokerSelected{BrokerId: "a broker"},
+						},
+					},
+					{
+						Data: &EventData_AuthModeSelected{
+							AuthModeSelected: &Events_AuthModeSelected{AuthModeId: "auth mode"},
+						},
+					},
+				},
+			},
+
+			wantError: "poll response data member 1 invalid: missing event type",
+		},
+
+		"Error pollResponse packet with event with missing type": {
+			gdmData: &Data{
+				Type: DataType_pollResponse,
+				PollResponse: []*EventData{
+					{},
+					{
+						Type: EventType_authModeSelected,
+						Data: &EventData_AuthModeSelected{
+							AuthModeSelected: &Events_AuthModeSelected{AuthModeId: "auth mode"},
+						},
+					},
+				},
+			},
+
+			wantError: "poll response data member 0 invalid: missing event type",
+		},
+
+		"Error pollResponse packet with unexpected data": {
+			gdmData: &Data{
+				Type:         DataType_pollResponse,
+				PollResponse: []*EventData{},
+				Event:        &EventData{},
+			},
+
+			wantError: "field Event should not be defined",
+		},
+	}
+
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			bytes, err := tc.gdmData.JSON()
+			if tc.wantError != "" {
+				require.ErrorContains(t, err, tc.wantError)
+				return
+			}
+
+			require.NoError(t, err)
```

Comment:
nitpick: Can you move the `require.NoError` one line above so it stays on the same "block" as where the error first got initialized?